### PR TITLE
Improve `add_feed_dict` docstrings

### PR DESCRIPTION
**Pull request by ElenaKhaustova**
## Description
Solves https://github.com/kedro-org/kedro/issues/3612

## Development notes

- Updated the docstrings for `DataCatalog`'s `add_feed_dict` method and extended them with an additional example.
- Created `form_dict` alias for `add_feed_dict` method and added a note to the docstring that the method will be renamed in the `0.20.0` release.

## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

**Comment by ElenaKhaustova** [comment]
I would also suggest replacing `io` with `catalog` in the rest of the `DataCatalog`'s docstrings as the last feels more intuitive.

File: kedro/io/data_catalog.py [hunk 1]
```diff
@@ -681,29 +681,41 @@ def add_all(
             self.add(name, dataset, replace)
 
     def add_feed_dict(self, feed_dict: dict[str, Any], replace: bool = False) -> None:
-        """Adds instances of ``MemoryDataset``, containing the data provided
-        through feed_dict.
+        """This function adds datasets to the ``DataCatalog`` using the data provided through the `feed_dict`.
```
Comment (astrojuanlu) [inline kedro/io/data_catalog.py:684]: I don't think we have a style guide for docstrings yet? In any case, following [Pydocstyle](http://www.pydocstyle.org/en/stable/error_codes.html) D404 and D401,

```suggestion
        """Add datasets to the ``DataCatalog`` using the data provided through the `feed_dict`.
```

File: kedro/io/data_catalog.py [hunk 2]
```diff
@@ -713,6 +725,9 @@ def add_feed_dict(self, feed_dict: dict[str, Any], replace: bool = False) -> Non
 
             self.add(dataset_name, dataset, replace)
 
+    # Alias for add_feed_dict method
+    from_dict = add_feed_dict
```
Comment (astrojuanlu) [inline kedro/io/data_catalog.py:729]: Pedantic nitpick, but when I read `from_dict` I assume it's a classmethod, like

```python
class DataCatalog:
  @classmethod
  def from_dict(cls, data_dict: dict[str, AbstractDataset]):
    ...
    return cls
```

So not sure if I'd add this alias.

(I know this comes from https://github.com/kedro-org/kedro/issues/3612#issuecomment-2034961020 and that I asserted, but my response there is incomplete)

File: kedro/io/data_catalog.py [hunk 3]
```diff
@@ -681,29 +681,41 @@ def add_all(
             self.add(name, dataset, replace)
 
     def add_feed_dict(self, feed_dict: dict[str, Any], replace: bool = False) -> None:
-        """Adds instances of ``MemoryDataset``, containing the data provided
-        through feed_dict.
+        """This function adds datasets to the ``DataCatalog`` using the data provided through the `feed_dict`.
+
+        `feed_dict` dictionary key is used as a dataset name, and a value is used to create an instance of
+        ``MemoryDataset`` before adding to the ``DataCatalog`` for all the value types except of ``AbstractDataset``.
+        In the last case, the ``AbstractDataset`` is added as it is.
```
Comment (astrojuanlu) [inline kedro/io/data_catalog.py:688]: What about

```suggestion
        `feed_dict` keys are used as dataset names, and values can either be raw data or instances of ``AbstractDataset``.
        In the former case, instances of ``MemoryDataset`` are automatically created before adding to the ``DataCatalog``.
```

?

File: kedro/io/data_catalog.py (hunk 2 above)
Comment (merelcht) [reply kedro/io/data_catalog.py:729]: This is actually quite a significant change. I think we should discuss first what a good alternative is, before just adding an alias.

File: kedro/io/data_catalog.py (hunk 1 above)
Comment (merelcht) [inline kedro/io/data_catalog.py:684]: I'm still missing a bit more explanation here about what `feed_dict` is. Adding that would make a huge difference to me.

Comment (datajoely) [reply kedro/io/data_catalog.py:684]: Add additional ?
```suggestion
        """Add additional datasets to the ``DataCatalog`` using the data provided through the `feed_dict`.
```

File: kedro/io/data_catalog.py (hunk 3 above)
Comment (datajoely) [reply kedro/io/data_catalog.py:688]: I actually don't think mentioning `AbstractDataset` is useful here, we shouldn't really have instances of an interface right? Does the user gain anything for understanding this implementation detail?

File: kedro/io/data_catalog.py (hunk 2 above)
Comment (datajoely) [reply kedro/io/data_catalog.py:729]: This has always been an undiscovered part of the API, I agree with @merelcht that we should think really hard / research what a better name should be.

Comment (ElenaKhaustova) [reply kedro/io/data_catalog.py]: I removed an alias until we are sure about adding it and agree on its name.

To @astrojuanlu's point, if we decide to keep an alias, we can use the `add_from_dict` name instead.

Or, as an alternative, we can add a new implementation (`add_from_dict` / `add_datasets`) and use it inside `add_feed_dict` with a deprecation warning.

File: kedro/io/data_catalog.py (hunk 1 above)
Comment (ElenaKhaustova) [reply kedro/io/data_catalog.py]: I didn't fully understand why we need `additional` here, as we can add them to the empty catalog as well. So, it looks like we do not need to specify any adding conditions.

File: kedro/io/data_catalog.py (hunk 3 above)
Comment (ElenaKhaustova) [reply kedro/io/data_catalog.py]: That's a good point. I rephrased it, but I still think it's useful to distinguish between the two cases when adding raw data and a dataset.

File: kedro/io/data_catalog.py (hunk 1 above)
Comment (ElenaKhaustova) [reply kedro/io/data_catalog.py]: Done

**Review by astrojuanlu — APPROVED** [review]
LGTM!

**Review by ankatiyar — APPROVED** [review]
much clearer!

**Review by merelcht — APPROVED** [review]
LGTM! ⭐

**Pull request merged**